File: backend/app/tools/content_cache.py
"""Content-addressed cache path definitions for verified download bytes."""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from pathlib import Path

_SHA256 = re.compile(r"^[0-9a-f]{64}$")


def _canonical_sha256(value: str) -> str:
    checksum = value.strip().lower()
    if not _SHA256.fullmatch(checksum):
        raise ValueError("SHA-256 must contain exactly 64 hexadecimal characters")
    return checksum


def canonical_request_hash(
    database: str,
    accession: str,
    url: str,
) -> str:
    """Compute a deterministic SHA-256 from canonical request identity.

    The cache key uses ``(database, accession, url)`` — never free-text
    keywords alone — so identical source requests resolve to the same
    cached blob without re-downloading.
    """
    canonical = json.dumps(
        {
            "database": database.strip().lower(),
            "accession": accession.strip().lower(),
            "url": url.strip(),
        },
        sort_keys=True,
        separators=(",", ":"),
    )
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


@dataclass(frozen=True)
class ContentCache:
    """Resolve cache paths without treating free-text queries as cache keys."""

    root: Path

    def blob_path(self, sha256: str) -> Path:
        checksum = _canonical_sha256(sha256)
        parent = self.root / "blobs" / "sha256" / checksum[:2] / checksum[2:4]
        parent.mkdir(parents=True, exist_ok=True)
        return parent / checksum

    def metadata_path(self, canonical_request_hash: str) -> Path:
        request_hash = _canonical_sha256(canonical_request_hash)
        parent = self.root / "metadata"
        parent.mkdir(parents=True, exist_ok=True)
        return parent / f"{request_hash}.json"

    def read_metadata(self, request_hash: str) -> dict[str, str] | None:
        """Return cached metadata for *request_hash* or ``None`` if absent."""
        path = self.metadata_path(request_hash)
        if not path.is_file():
            return None
        try:
            data = json.loads(path.read_text("utf-8"))
        except (json.JSONDecodeError, OSError):
            return None
        if not isinstance(data, dict) or "sha256" not in data:
            return None
        return data

    def write_metadata(self, request_hash: str, metadata: dict[str, str]) -> None:
        """Persist *metadata* for *request_hash* to the metadata directory."""
        path = self.metadata_path(request_hash)
        path.write_text(json.dumps(metadata, sort_keys=True) + "\n", "utf-8")

File: server/src/external/acquisition/content-cache.ts
/**
 * Content-addressed cache path definitions (Python
 * ``app/tools/content_cache.py`` parity).
 */

import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";

const SHA256_RE = /^[0-9a-f]{64}$/;

function canonicalSha256(value: string): string {
  const checksum = value.trim().toLowerCase();
  if (!SHA256_RE.test(checksum)) {
    throw new TypeError("SHA-256 must contain exactly 64 hexadecimal characters");
  }
  return checksum;
}

/**
 * Non-default HTTP request identity for the content cache. GET requests
 * without a body keep the historical (database, accession, url) key; POST
 * acquisitions include method + body so distinct requests to the same URL
 * never share a cache entry.
 */
export interface RequestIdentityVariant {
  method: string;
  body?: string;
}

/** Deterministic SHA-256 of canonical request identity (database, accession, url). */
export function canonicalRequestHash(
  database: string,
  accession: string,
  url: string,
  variant?: RequestIdentityVariant,
): string {
  // Key order matches Python json.dumps(..., sort_keys=True) for the base
  // (GET) identity: accession, database, url. The POST method/body extension
  // is TypeScript-only (Python had no POST acquisition path) and is appended
  // in insertion order after the base keys.
  const canonical = JSON.stringify({
    accession: accession.trim().toLowerCase(),
    database: database.trim().toLowerCase(),
    url: url.trim(),
    ...(variant === undefined ? {} : {
      method: variant.method,
      body: variant.body ?? "",
    }),
  });
  return createHash("sha256").update(canonical).digest("hex");
}

export interface ContentCacheRoot {
  root: string;
}

export interface CachedMetadata {
  sha256: string;
  size_bytes?: string;
  media_type?: string;
  [key: string]: string | undefined;
}

export class ContentCache {
  readonly root: string;

  constructor(root: string) {
    this.root = path.resolve(root);
  }

  blobPath(sha256: string): string {
    const checksum = canonicalSha256(sha256);
    const parent = path.join(this.root, "blobs", "sha256", checksum.slice(0, 2), checksum.slice(2, 4));
    return path.join(parent, checksum);
  }

  metadataPath(requestHash: string): string {
    const hash = canonicalSha256(requestHash);
    return path.join(this.root, "metadata", `${hash}.json`);
  }

  async readMetadata(requestHash: string): Promise<CachedMetadata | null> {
    const file = this.metadataPath(requestHash);
    let data: unknown;
    try {
      data = JSON.parse(await readFile(file, "utf8"));
    } catch {
      return null;
    }
    if (typeof data !== "object" || data === null || !("sha256" in data) || typeof (data as { sha256: unknown }).sha256 !== "string") {
      return null;
    }
    return data as CachedMetadata;
  }

  async writeMetadata(requestHash: string, metadata: Record<string, string>): Promise<void> {
    const file = this.metadataPath(requestHash);
    await mkdir(path.dirname(file), { recursive: true });
    const ordered = Object.fromEntries(Object.entries(metadata).sort(([left], [right]) => left.localeCompare(right)));
    await writeFile(file, `${JSON.stringify(ordered)}\n`, "utf8");
  }
}
